[WebAssembly SIMD] Support integer and floating-point arithmetic on Intel

ddegazio opened the pull request:
#### 155740a50205d47608751bd4745d562202bca427
<pre>
[WebAssembly SIMD] Support integer and floating-point arithmetic on Intel
<a href="https://bugs.webkit.org/show_bug.cgi?id=248549">https://bugs.webkit.org/show_bug.cgi?id=248549</a>

Reviewed by Justin Michaud.

Adds support for basic SIMD integer arithmetic and floating-point arithmetic
instructions on Intel.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorAdd):
(JSC::MacroAssemblerX86_64::vectorSub):
(JSC::MacroAssemblerX86_64::vectorMul):
(JSC::MacroAssemblerX86_64::vectorDiv):
(JSC::MacroAssemblerX86_64::vectorCeil):
(JSC::MacroAssemblerX86_64::vectorFloor):
(JSC::MacroAssemblerX86_64::vectorTrunc):
(JSC::MacroAssemblerX86_64::vectorNearest):
(JSC::MacroAssemblerX86_64::vectorDotProductInt32):
(JSC::MacroAssemblerX86_64::vectorNeg): Deleted.
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vectorFadd):
(JSC::X86Assembler::vectorAdd):
(JSC::X86Assembler::vectorFsub):
(JSC::X86Assembler::vectorSub):
(JSC::X86Assembler::vectorFmul):
(JSC::X86Assembler::vectorMul):
(JSC::X86Assembler::vectorFdiv):
(JSC::X86Assembler::vroundps_rr):
(JSC::X86Assembler::vroundpd_rr):
(JSC::X86Assembler::vpmaddwd_rrr):
(JSC::X86Assembler::X86InstructionFormatter::SingleInstructionBufferWriter::memoryModRM):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addSIMDV_V):

Canonical link: <a href="https://commits.webkit.org/257511@main">https://commits.webkit.org/257511@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/61f605b9f0c4ff6118eeabecdc80e8f8b6ebeb83

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99206 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8418 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/32339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108599 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/168852 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/8970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/85738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/91711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/106536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104964 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/8970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/32339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/91711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/76/builds/8970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/32339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/91711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89914 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/2298 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/32339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85745 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2189 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/85738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28863 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/7242 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/32339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88609 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2628 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/3667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19828 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -548,6 +547,19 @@ class AirIRGenerator {
         AIR_OP_CASE(TruncSat)
+#if CPU(X86_64)
+        else if (op == SIMDLaneOperation::Neg) {
+            // x86_64 has no direct vector negate, so we expand vxv.neg v into vxv.sub 0, v
```
We can probably just leave this at "has no vector negate instruction"

justinmichaud commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2416,6 +2439,143 @@ class X86Assembler {
+            m_formatter.vexNdsLigWigThreeByteOp(PRE_OPERAND_SIZE, VexImpliedBytes::ThreeBytesOp38, OP3_PMULLD, (RegisterID)dest, (RegisterID)left, (RegisterID)right);
+            break;
+        case SIMDLane::i64x2:
+            ASSERT_NOT_REACHED("TODO: Implement i64x2.mul. It's not natively supported on Intel so we need to do some weird things.");
```
release assert not reached, and mention that it should have been lowered already

ddegazio replied:
I've switched the asserts to release asserts, and added a more informative message in the latest version.

justinmichaud reviewed (approved):
r=me

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86Common.cpp:
```diff
@@ -810,18 +810,23 @@ void MacroAssemblerX86Common::collectCPUFeatures()
             s_sse4_2CheckState = (cpuid[2] & (1 << 20)) ? CPUIDCheckState::Set : CPUIDCheckState::Clear;
             s_popcntCheckState = (cpuid[2] & (1 << 23)) ? CPUIDCheckState::Set : CPUIDCheckState::Clear;
             s_avxCheckState = (cpuid[2] & (1 << 28)) ? CPUIDCheckState::Set : CPUIDCheckState::Clear;
+            s_avxCheckState = (cpuid[2] & (1 << 28)) ? CPUIDCheckState::Set : CPUIDCheckState::Clear;
```
This line is not necessary, and not correct (since CPUID is done with 0x1).

ddegazio replied:
Oh, good catch! I think that's a typo, I'll remove the second line.

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86Common.cpp:
```diff
@@ -810,18 +810,23 @@ void MacroAssemblerX86Common::collectCPUFeatures()
             size_t valSize = sizeof(val);
             int rc = sysctlbyname("hw.optional.bmi1", &val, &valSize, nullptr, 0);
             s_bmi1CheckState = (rc >= 0 && val) ? CPUIDCheckState::Set : CPUIDCheckState::Clear;
+
+            rc = sysctlbyname("hw.optional.avx2_0", &val, &valSize, nullptr, 0);
+            s_avx2CheckState = (rc >= 0 && val) ? CPUIDCheckState::Set : CPUIDCheckState::Clear;
```
Looks good. We are not using `getCPUID(0x7)` because it slows down hypervisors (VMExit happens).

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerX86Common.cpp:
```diff
@@ -810,18 +810,23 @@ void MacroAssemblerX86Common::collectCPUFeatures()
         {
             CPUID cpuid = getCPUID(0x7);
             s_bmi1CheckState = (cpuid[2] & (1 << 3)) ? CPUIDCheckState::Set : CPUIDCheckState::Clear;
+            s_avx2CheckState = (cpuid[2] & (1 << 5)) ? CPUIDCheckState::Set : CPUIDCheckState::Clear;
```
Looks good.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2590,6 +2613,143 @@ class X86Assembler {
         m_formatter.immediate8((uint8_t)controlBits);
     }
 
+    void vectorFadd(SIMDLane lane, XMMRegisterID left, XMMRegisterID right, XMMRegisterID dest)
```
Let's name them with more direct name. This `vectorFadd` abstraction should be done in the MacroAssembler, and X86Assembler should be one-on-one correspondence with the actual X86 instructions.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2590,6 +2613,143 @@ class X86Assembler {
+        }
+    }
+
+    void vectorAdd(SIMDLane lane, XMMRegisterID left, XMMRegisterID right, XMMRegisterID dest)
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2590,6 +2613,143 @@ class X86Assembler {
+        }
+    }
+
+    void vectorFsub(SIMDLane lane, XMMRegisterID left, XMMRegisterID right, XMMRegisterID dest)
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2590,6 +2613,143 @@ class X86Assembler {
+            m_formatter.vexNdsLigWigCommutativeTwoByteOp((OneByteOpcodeID)0, OP2_ADDPS_VpsWps, (RegisterID)dest, (RegisterID)left, (RegisterID)right);
+            break;
+        case SIMDLane::f64x2:
+            m_formatter.vexNdsLigWigCommutativeTwoByteOp(PRE_OPERAND_SIZE, OP2_ADDPD_VpdWpd, (RegisterID)dest, (RegisterID)left, (RegisterID)right);
```
So, let's add `addps_rr`, `addpd_rr` etc.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2590,6 +2613,143 @@ class X86Assembler {
+        }
+    }
+
+    void vectorSub(SIMDLane lane, XMMRegisterID left, XMMRegisterID right, XMMRegisterID dest)
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2590,6 +2613,143 @@ class X86Assembler {
+        }
+    }
+
+    void vectorFmul(SIMDLane lane, XMMRegisterID left, XMMRegisterID right, XMMRegisterID dest)
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2590,6 +2613,143 @@ class X86Assembler {
+        }
+    }
+
+    void vectorMul(SIMDLane lane, XMMRegisterID left, XMMRegisterID right, XMMRegisterID dest)
```
Ditto.

Constellation commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -2590,6 +2613,143 @@ class X86Assembler {
+        }
+    }
+
+    void vectorFdiv(SIMDLane lane, XMMRegisterID left, XMMRegisterID right, XMMRegisterID dest)
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -549,6 +548,19 @@ class AirIRGenerator {
         AIR_OP_CASE(ExtendHigh)
         AIR_OP_CASE(ExtendLow)
         AIR_OP_CASE(TruncSat)
+#if CPU(X86_64)
+        else if (op == SIMDLaneOperation::Neg) {
+            // x86_64 has no vector negate instruction, so we expand vxv.neg v into vxv.sub 0, v
+            // here to give B3/Air a chance to optimize out repeated vector zeroing.
+            TypedTmp zero = addConstant(v128_t());
+            result = tmpForType(Types::V128);
+            append(VectorSub, zero, v, result);
+            return { };
+        }
+#else
+        AIR_OP_CASE(Neg)
+#endif
```
Or, you can lower this in `AirLowerMacros.cpp` for X86, then we can keep WasmAirIRGenerator.cpp simpler.

ddegazio replied:
We're doing it in the AirIRGenerator in the hopes that by expanding negates as early as possible, we give the backend as many chances as possible to optimize out repeated uses of the constant. Probably not very meaningful for Air, but it may matter more in B3IRGenerator (where Air macros would expand after earlier optimization) so our plan is to keep it here so the two IRGenerators are consistent.

Constellation commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -627,6 +639,25 @@ class AirIRGenerator {
+            append(B3::Air::Mul64, first, second, second);
+            addReplaceLane(info, 1, result, second, result);
+            return { };
+        }
```
Ditto.

ddegazio replied:
This one can probably be an Air lower macro, I'll move it.

Constellation reviewed (commented):
Commented.

ddegazio committed:
[WebAssembly SIMD] Support integer and floating-point arithmetic on Intel
https://bugs.webkit.org/show_bug.cgi?id=248549

Reviewed by Justin Michaud.

Adds support for basic SIMD integer arithmetic and floating-point arithmetic
instructions on Intel.

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorAdd):
(JSC::MacroAssemblerX86_64::vectorSub):
(JSC::MacroAssemblerX86_64::vectorMul):
(JSC::MacroAssemblerX86_64::vectorDiv):
(JSC::MacroAssemblerX86_64::vectorCeil):
(JSC::MacroAssemblerX86_64::vectorFloor):
(JSC::MacroAssemblerX86_64::vectorTrunc):
(JSC::MacroAssemblerX86_64::vectorNearest):
(JSC::MacroAssemblerX86_64::vectorDotProductInt32):
(JSC::MacroAssemblerX86_64::vectorNeg): Deleted.
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::vectorFadd):
(JSC::X86Assembler::vectorAdd):
(JSC::X86Assembler::vectorFsub):
(JSC::X86Assembler::vectorSub):
(JSC::X86Assembler::vectorFmul):
(JSC::X86Assembler::vectorMul):
(JSC::X86Assembler::vectorFdiv):
(JSC::X86Assembler::vroundps_rr):
(JSC::X86Assembler::vroundpd_rr):
(JSC::X86Assembler::vpmaddwd_rrr):
(JSC::X86Assembler::X86InstructionFormatter::SingleInstructionBufferWriter::memoryModRM):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addSIMDV_V):

Canonical link: https://commits.webkit.org/257511@main
